core/interop/src/lib.rs

```diff
@@ -39,67 +38,255 @@ impl<T: IntoIterator<Item = (JsString, NativeFunction)> + Clone> IntoJsModule fo
 /// This trait does not require the implementing type to be `Copy`, which
 /// can lead to undefined behaviour if it contains Garbage Collected objects.
 ///
+/// This trait is implemented for functions with various signatures.
+///
+/// For example:
+/// ```
+/// # use boa_engine::{Context, JsValue, NativeFunction};
+/// # use boa_interop::IntoJsFunctionUnsafe;
+/// # let mut context = Context::default();
+/// let f = |a: i32, b: i32| a + b;
+/// let f = unsafe { f.into_js_function_unsafe(&mut context) };
+/// let result = f.call(&JsValue::undefined(), &[JsValue::from(1), JsValue::from(2)], &mut context).unwrap();
+/// assert_eq!(result, JsValue::new(3));
+/// ```
+///
+/// Since the `IntoJsFunctionUnsafe` trait is implemented for `FnMut`, you can
+/// also use closures directly:
+/// ```
+/// # use boa_engine::{Context, JsValue, NativeFunction};
+/// # use boa_interop::IntoJsFunctionUnsafe;
+/// # use std::cell::RefCell;
+/// # use std::rc::Rc;
+/// # let mut context = Context::default();
+/// let mut x = Rc::new(RefCell::new(0));
+/// // Because NativeFunction takes ownership of the closure,
+/// // the compiler cannot be certain it won't outlive `x`, so
+/// // we need to create a `Rc<RefCell>` and share it.
+/// let f = unsafe {
+///     let x = x.clone();
+///     move |a: i32| *x.borrow_mut() += a
+/// };
+/// let f = unsafe { f.into_js_function_unsafe(&mut context) };
+/// f.call(&JsValue::undefined(), &[JsValue::from(1)], &mut context).unwrap();
+/// f.call(&JsValue::undefined(), &[JsValue::from(4)], &mut context).unwrap();
+/// assert_eq!(*x.borrow(), 5);
+/// ```
+///
 /// # Safety
 /// For this trait to be implemented safely, the implementing type must not contain any
 /// garbage collected objects (from [`boa_gc`]).
-pub unsafe trait IntoJsFunctionUnsafe {
+pub unsafe trait IntoJsFunctionUnsafe<Args, Ret> {
```

Comment:
I think we can seal both `IntoJsFunction*` traits to remove the `unsafe` marker on it. Those will be provided by us anyways. That would also allow us to rename them to `IntoJsFunction` and `IntoJsFunctionCopied`, which is a bit more explanatory.

Reply:
Good idea, done.

Reply:
> That would also allow us to rename them to IntoJsFunction and IntoJsFunctionCopied, which is a bit more explanatory.

I'd like to keep the `Unsafe` suffix on the unsafe one though. Keeps users on their toes :)

Reply:
That's also fine. Let's use `UnsafeIntoJsFunction` and `IntoJsFunction` then.

Reply:
Done. PTAL